Method for saving image into a Data BLOB

vojtamolda opened the pull request:
Hello @t-ae ,

I implemented a change that adds a method for encoding an image to a specific format purely in-memory, without writing a temporary file. Signature of the method is `Image.inscribe(format:)` and it returns the image in the given format as a `Data` BLOB.

To explain it better here's an example use of the new method:
```swift
let image = Image<RGB, UInt8>(width: 32, height: 32, color: .red)
let data = try image.inscribe(format: .jpeg(quality: 80))
```

PS: There's also a few documentation improvements and typo fixes sprinkled throughout the commits...


### Motivation

I believe the feature has many more use cases but I ended up needing it in order to create two specific things. 

First, a video from a collection of `Image` frames. It's much faster to encode the video purely in-memory without storing any temporary files. Especially for longer videos there's also a significant disk space saving. The solution I used in my case is to pipe the individual frames to FFmpeg subprocess. FFmpeg encodes the video on the fly and pipes the chunks back to the main application via stdout.

Second, I wanted to display the `Image` in Colab/Jupyter notebook. The protocol requires the image to be transported in a JSON-able format. So the raw bytes need to be base64 encoded. Again it is easier and faster without a temporary file.


### Naming

Naming is hard... Any suggestions for improvements are welcome. Besides the `Image.inscribe(...)` I also considered using `Image.data(...) signature`. Personally, I like the `.data(...)` option a bit more. It is very similar to what people are used from the Foundation library. Specifically, when one wants binary `Data` representation of a `String` then the method that does the job is `String.data(using:)`.

Unfortunately, `Image.data` is already taken for actual pixels. Even though it's not declared as a public property it is used almost everywhere throughout the codebase. So renaming it to something else just to make room for my small improvement isn't a good idea.


### Implementation

Under the hood, `Image.inscribe(...)` uses the `stbi_write_*_to_func(...)` family of C functions. These functions accept a pointer to a writing callback instead of the filename. Thanks to the interoperability with C, the writing callback can be implemented in Swift. The closure keeps receiving raw byte chunks and gradually builds the `Data` BLOB buffer that eventually holds the entire image encoded in the desired file format.

vojtamolda committed:
Method for inscribing image into a Data BLOB

vojtamolda committed:
Unit tests for accessing image as in-memory Data BLOB

vojtamolda committed:
Add example of image encoded to a Data BLOB to readme

t-ae commented on Sources/Swim/IO/InscribeImage.swift:
```diff
@@ -0,0 +1,110 @@
+
+@inlinable
+func inscribe(context: UnsafeMutableRawPointer?, data: UnsafeMutableRawPointer?, size: Int32) -> Void {
+    guard data != nil &&  context != nil else { return }
```
Extra white space.
```suggestion
    guard data != nil && context != nil else { return }
```

t-ae commented:
Great! I didn't know this can be done in Swift side!

About naming: I regret I chose the name `data` for pixel data. `data` is definitely better.
By the way I checked `UIImage` and found it has [`pngData`](https://developer.apple.com/documentation/uikit/uiimage/1624096-pngdata) and [`jpegData`](https://developer.apple.com/documentation/uikit/uiimage/1624115-jpegdata) for this purpose.
What do you think of having `pngData`, `jpegData`, `bmpData` instead of `inscribe`?
I think they are more integrated and friendly to Apple platform developers.

vojtamolda commented:
Checking how `UIImage` does things is definitely a good idea that didn't occur to me. I read through the `jpeg`/`pngData` doc links but I can't say I'm a big fan of this approach. It's difficult to put it into words but using multiple methods where one can do the job doesn't feel right to me. I'm not a history expert but it seems that these functions originate way back from the Objective-C heydays.

The `jpeg`/`pngData` methods would also look strange next to the `Image.write(...)` call. The write function uses a nice `WriteFormat` enum to specify the file format. There's a case for each supported format with an additional quality property for JPEGs. I personally like this enum a lot. It feels to me that using a single method with a `WriteFormat` argument is a better fit to the rest of the API and Swift in general.

So I'd like to suggest new and hopefully better fitting name for `Image.incsribe(...)`. Here is a few ideas together with an example to show how they would look from the point of use:

1. **`rawData(format: WriteFormat) throws -> Data`**
    ```swift
    let data = try? image.rawData(format: .png)
    ```
2. **`formatData(as format: WriteFormat) throws -> Data`**
    ```swift
    let data = try? image.formatData(as: .png)
    ```
3. **`format(as format: WriteFormat) throws -> Data`**
    ```swift
    let data = try? image.format(as: .png)
    ```

What do you think? Do you think any of those would work? Or if you really like the `jpeg`/`pngData` and the compatibility with `UIImage` as you can see I don't have particularly strong argument against it. It's mostly gut feel...

t-ae commented:
> I read through the jpeg/pngData doc links but I can't say I'm a big fan of this approach. It's difficult to put it into words but using multiple methods where one can do the job doesn't feel right to me. 

I also suspected so. We already need 3 methods, and will need new method for each newly supported-format.
 
> I'm not a history expert but it seems that these functions originate way back from the Objective-C heydays.

They were added in Swift 4.2😂 But, yes, they stemed from `UIImagePNGRepresentation` and `UIImageJPEGRepresentation`. I don't know why Apple developer didn't put them together into a single method.

> So I'd like to suggest new and hopefully better fitting name for Image.incsribe(...). Here is a few ideas together with an example to show how they would look from the point of use:

I like `formatData(as format: WriteFormat) throws -> Data`. This partially corresponds to `pngData`/`jpegData`, and is less ambiguous than `rawData` (It can sound like raw pixel data).
But, according to API Design Guideline, 

> Name functions and methods according to their side-effects
> Those without side-effects should read as noun phrases, e.g. x.distance(to: y), i.successor().
> Those with side-effects should read as imperative verb phrases, e.g., print(x), x.sort(), x.append(y).

https://swift.org/documentation/api-design-guidelines/#naming

`inscribe` doesn't have side-effect, so it should be a noun phrase.
`formatData` is also a verb phrase. Adopting the guideline it should be `formattedData`.

Another idea came to my mind is `fileData`. It sounds some clearer than `formattedData`.
What do you think about them?

vojtamolda committed:
Rename Image.inscribe method to fileData

vojtamolda commented:
Thanks a lot for the comment, @t-ae. I agree with all your points and the Swift 4.2 addition made me literally laugh out loud.

I like `fileData(format:)` the most. It also suggests that the returned BLOB matches the file content. I have just updated the PR to the new name and also removed the extra white space.

t-ae commented:
Thank you @vojtamolda.
I'll do some work tomorrow and then bump minor version.

vojtamolda commented:
Awesome. Thanks!

t-ae commented:
I added new initializers that takes image file `Data` and pushed new version 3.9.0.
Again, thank you for your contribution @vojtamolda!
